src/services/signer_service.ts

```diff
@@ -0,0 +1,183 @@
+    public static isEligibleForFreeMetaTxn(apiKey: string): boolean {
+        return WHITELISTED_API_KEYS_META_TXN_FILLS.includes(apiKey);
+    }
+
+    private static _createWeb3Provider(
+        rpcHost: string,
+        privateWalletSubprovider: PrivateKeyWalletSubprovider,
+        nonceTrackerSubprovider: NonceTrackerSubprovider,
+    ): SupportedProvider {
+        const WEB3_RPC_RETRY_COUNT = 3;
+        const providerEngine = new Web3ProviderEngine();
+        providerEngine.addProvider(nonceTrackerSubprovider);
+        providerEngine.addProvider(privateWalletSubprovider);
+        const rpcSubproviders = SignerService._range(WEB3_RPC_RETRY_COUNT).map(
+            (_index: number) => new RPCSubprovider(rpcHost),
+        );
+        providerEngine.addProvider(new RedundantSubprovider(rpcSubproviders));
+        providerUtils.startProviderEngine(providerEngine);
+        return providerEngine;
+    }
+    private static _range(rangeCount: number): number[] {
+        return [...Array(rangeCount).keys()];
+    }
+    private static _calculateProtocolFee(numOrders: number, gasPrice: BigNumber): BigNumber {
+        return new BigNumber(150000).times(gasPrice).times(numOrders);
+    }
+    constructor() {
+        this._privateWalletSubprovider = new PrivateKeyWalletSubprovider(META_TXN_RELAY_PRIVATE_KEY);
+        this._nonceTrackerSubprovider = new NonceTrackerSubprovider();
+        this._provider = SignerService._createWeb3Provider(
+            ETHEREUM_RPC_URL,
+            this._privateWalletSubprovider,
+            this._nonceTrackerSubprovider,
+        );
+        this._contractWrappers = new ContractWrappers(this._provider, { chainId: CHAIN_ID });
+        this._web3Wrapper = new Web3Wrapper(this._provider);
+    }
+    public async validateZeroExTransactionFillAsync(
+        zeroExTransaction: ZeroExTransactionWithoutDomain,
+        signature: string,
+    ): Promise<BigNumber> {
+        // Verify 0x txn won't expire in next 60 seconds
+        // tslint:disable-next-line:custom-no-magic-numbers
+        const sixtySecondsFromNow = new BigNumber(+new Date() + 60);
+        if (zeroExTransaction.expirationTimeSeconds <= sixtySecondsFromNow) {
+            throw new Error('zeroExTransaction expirationTimeSeconds in less than 60 seconds from now');
+        }
+
+        const devUtils = new DevUtilsContract(this._contractWrappers.contractAddresses.devUtils, this._provider);
+        const decodedArray = await devUtils.decodeZeroExTransactionData(zeroExTransaction.data).callAsync();
+        const orders = decodedArray[1];
+
+        // Verify orders don't expire in next 60 seconds
+        orders.forEach(order => {
+            if (order.expirationTimeSeconds <= sixtySecondsFromNow) {
+                throw new Error('Order included in zeroExTransaction expires in less than 60 seconds from now');
+            }
+        });
+
+        const gasPrice = zeroExTransaction.gasPrice;
```

Comment:
We should have some acceptable limit of gasPrice

Reply:
Do you mean an upper limit?

Reply:
Yes, as in should we validate this in terms of a gas oracle (ethgasstation for example) so its not submitted at 200 gwei when fast is 8 gwei.

Reply:
Added in a check to make sure gasPrice isn't more than 3x ethGasStation fast.